honeybee_grasshopper_core/src/HB Apertures by Ratio.py

```diff
@@ -0,0 +1,209 @@
+            are input here, different separation distances will be assigned based
+            on cardinal direction, starting with north and moving clockwise.
+            Default: 0.
+        operable_: An optional boolean to note whether the generated Apertures
+            can be opened for ventilation. If an array of booleans are input
+            here, different operable properties will be assigned based on
+            cardinal direction, starting with north and moving clockwise.
+            Default: False.
+        ep_constr_: Optional text for an energy construction to be used for all
+            generated apertures. This text will be used to look up a construction
+            in the window construction library. This can also be a custom
+            WindowConstruction object. If an array of text or construction objects
+            are input here, different constructions will be assigned based on
+            cardinal direction, starting with north and moving clockwise.
+        rad_mat_: Optional text for a radiance material to be used for all
+            generated apertures. This text will be used to look up a material
+            in the material library. This can also be a custom material object.
+            If an array of text or material objects are input here, different
+            materials will be assigned based on cardinal direction, starting
+            with north and moving clockwise.
+        _run: Set to True to run the component.
+    
+    Returns:
+        report: Reports, errors, warnings, etc.
+        hb_obj: The input Honeybee Face or Room with Apertures generated from
+            the input parameters.
+"""
+
+ghenv.Component.Name = "HB Apertures by Ratio"
+ghenv.Component.NickName = 'AperturesByRatio'
+ghenv.Component.Message = '0.1.0\nOCT_28_2019'
+ghenv.Component.Category = "HoneybeeCore"
+ghenv.Component.SubCategory = '0 :: Create'
+ghenv.Component.AdditionalHelpFromDocStrings = "4"
+
+try:  # import the core honeybee dependencies
+    from honeybee.boundarycondition import Outdoors
+    from honeybee.facetype import Wall
+    from honeybee.room import Room
+    from honeybee.face import Face
+    from ladybug_rhino.config import tolerance, conversion_factor
+    from ladybug_rhino.grasshopper import all_required_inputs
+except ImportError as e:
+    raise ImportError('\nFailed to import honeybee:\n\t{}'.format(e))
+
+try:  # import the honeybee-energy dependencies
+    from honeybee_energy.lib.constructions import window_construction_by_name
+    from honeybee_energy.construction.window import WindowConstruction
+except ImportError:
+    pass  # honeybee-energy is not installed and ep_constr_ will not be avaialble
+
+
+def can_host_apeture(face):
+    """Test if a face is intended to host apertures (according to this component)."""
+    return isinstance(face.boundary_condition, Outdoors) and \
+        isinstance(face.type, Wall)
```

Comment:
Ok, I will add some text to the component documentation.

And I don't know what your second comment means. `type` is a property of `Face` itself.

Reply:
> And I don't know what your second comment means. type is a property of Face itself.

I said that about `can_host_aperture`.

```python
@property 
def can_host_aperture(self):
    """Test if a face is intended to host apertures (according to this component)."""
    return isinstance(self.boundary_condition, Outdoors) and \
        isinstance(selftype, Wall)
```

Reply:
Ah. I see what you mean. I originally didn't do this because this criteria is kinda specific to this component but I'm open to adding it as a property in the core.

The property would definitely need a different name, though, since we don't want to imply that users can't add apertures to Roofs/Floors or indoor Faces.  Does  `is_outdoor_wall` sound like a good name?

Reply:
So can you help me to understand why we need this check if you can add aperture to walls as well as roofs?

Reply:
Never mind. This is for the by ratio component. Disregard my comment about adding it as a property too.